core/server/master/src/main/java/alluxio/master/meta/DefaultMetaMaster.java

```diff
@@ -0,0 +1,405 @@
+import java.net.InetSocketAddress;
+import java.time.Clock;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * The master that handles all metadata management.
+ */
+@NotThreadSafe
+public final class DefaultMetaMaster extends AbstractMaster implements MetaMaster {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultMetaMaster.class);
+  private static final Set<Class<? extends Server>> DEPS =
+      ImmutableSet.<Class<? extends Server>>of(BlockMaster.class);
+
+  /** Indexes to help find master information in indexed set. */
+  private static final IndexDefinition<MasterInfo> ID_INDEX =
+      new IndexDefinition<MasterInfo>(true) {
+        @Override
+        public Object getFieldValue(MasterInfo o) {
+          return o.getId();
+        }
+      };
+  private static final IndexDefinition<MasterInfo> HOSTNAME_INDEX =
+      new IndexDefinition<MasterInfo>(true) {
+        @Override
+        public Object getFieldValue(MasterInfo o) {
+          return o.getHostname();
+        }
+      };
+
+  /** Handle to the block master. */
+  private final BlockMaster mBlockMaster;
+
+  /** The clock to use for determining the time. */
+  private final Clock mClock = new SystemClock();
+
+  /** The connect address for the rpc server. */
+  private final InetSocketAddress mRpcConnectAddress
+      = NetworkAddressUtils.getConnectAddress(NetworkAddressUtils.ServiceType.MASTER_RPC);
+
+  /** is true if the master is in safe mode. */
+  private boolean mIsInSafeMode;
+
+  /** Keeps track of standby masters which are in communication with the leader master. */
+  private final IndexedSet<MasterInfo> mMasters =
+      new IndexedSet<>(ID_INDEX, HOSTNAME_INDEX);
+  /** Keeps track of standby masters which are no longer in communication with the leader master. */
+  private final IndexedSet<MasterInfo> mLostMasters =
+      new IndexedSet<>(ID_INDEX, HOSTNAME_INDEX);
+
+  /** The master configuration checker. */
+  private final ServerConfigurationChecker mMasterConfigChecker = new ServerConfigurationChecker();
+  /** The worker configuration checker. */
+  private final ServerConfigurationChecker mWorkerConfigChecker = new ServerConfigurationChecker();
+
+  /** The hostname of this master. */
+  private String mMasterHostname = Configuration.get(PropertyKey.MASTER_HOSTNAME);
+
+  /** The master ID for this master. */
+  private AtomicReference<Long> mMasterId = new AtomicReference<>(-1L);
+
+  /** Client for all meta master communication. */
+  private final MetaMasterMasterClient mMetaMasterClient
+      = new MetaMasterMasterClient(MasterClientConfig.defaults());
+
+  /** The start time for when the master started serving the RPC server. */
```

Comment:
this is an unconventional definition of start time - usually start time would mean when the master process was started. The rpc server can be started and stopped many times over the lifetime of the master. Is there a reason to prefer to use the start time of the rpc server?

Reply:
Do not init and save StartTimeMs in MetaMaster anymore. Add callback function to get this info for MetaMasterClientServiceHandler.

Reply:
My guess is that we added the calculation of startTimeMs before we implemented starting/stopping, so the original intention was really to be closer to process uptime. @gpang do you know if we intended the master uptime to reset every time the rpc server starts? Now that we're going to show information about secondary master uptimes in the UI, I think it's important to make the uptime less confusing. Otherwise users will start all their masters at the same time, then be confused about why some masters end up with updates that are days apart even though none of them have been restarted.

For implementation, adding to `MasterContext` sgtm. We can also use that to share the rpc connect address